File: src/cpp/tests/test_motion.cpp
#include <gtest/gtest.h>

#include <memory>

#include "bcc950/constants.hpp"
#include "bcc950/motion.hpp"
#include "bcc950/position.hpp"
#include "mock_v4l2_device.hpp"

namespace bcc950 {
namespace {

/// Fixture providing a MotionController wired to a MockV4L2Device.
class MotionTest : public ::testing::Test {
protected:
    void SetUp() override {
        mock_ = std::make_unique<testing::MockV4L2Device>();
        position_ = PositionTracker{};
        motion_ = std::make_unique<MotionController>(mock_.get(), &position_);
    }

    std::unique_ptr<testing::MockV4L2Device> mock_;
    PositionTracker position_;
    std::unique_ptr<MotionController> motion_;
};

// ---- Pan tests ----

TEST_F(MotionTest, PanSetsSpeedAndStops) {
    motion_->pan(-1, 0.01);  // pan left, short duration

    const auto& calls = mock_->get_calls();
    ASSERT_GE(calls.size(), 2u);

    // First call: set pan speed to -1
    EXPECT_EQ(calls[0].first, CTRL_PAN_SPEED);
    EXPECT_EQ(calls[0].second, -1);

    // Second call: stop pan (set to 0)
    EXPECT_EQ(calls[1].first, CTRL_PAN_SPEED);
    EXPECT_EQ(calls[1].second, 0);
}

TEST_F(MotionTest, PanRightSetsPositiveSpeed) {
    motion_->pan(1, 0.01);

    const auto& calls = mock_->get_calls();
    ASSERT_GE(calls.size(), 2u);

    EXPECT_EQ(calls[0].first, CTRL_PAN_SPEED);
    EXPECT_EQ(calls[0].second, 1);

    EXPECT_EQ(calls[1].first, CTRL_PAN_SPEED);
    EXPECT_EQ(calls[1].second, 0);
}

TEST_F(MotionTest, PanUpdatesPositionTracker) {
    position_.pan = 0.0;
    motion_->pan(1, 0.5);

    // Position should have accumulated speed * duration
    EXPECT_DOUBLE_EQ(position_.pan, 0.5);
}

// ---- Tilt tests ----

TEST_F(MotionTest, TiltSetsSpeedAndStops) {
    motion_->tilt(1, 0.01);  // tilt up

    const auto& calls = mock_->get_calls();
    ASSERT_GE(calls.size(), 2u);

    EXPECT_EQ(calls[0].first, CTRL_TILT_SPEED);
    EXPECT_EQ(calls[0].second, 1);

    EXPECT_EQ(calls[1].first, CTRL_TILT_SPEED);
    EXPECT_EQ(calls[1].second, 0);
}

TEST_F(MotionTest, TiltDownSetsNegativeSpeed) {
    motion_->tilt(-1, 0.01);

    const auto& calls = mock_->get_calls();
    ASSERT_GE(calls.size(), 2u);

    EXPECT_EQ(calls[0].first, CTRL_TILT_SPEED);
    EXPECT_EQ(calls[0].second, -1);

    EXPECT_EQ(calls[1].first, CTRL_TILT_SPEED);
    EXPECT_EQ(calls[1].second, 0);
}

TEST_F(MotionTest, TiltUpdatesPositionTracker) {
    position_.tilt = 0.0;
    motion_->tilt(-1, 0.3);

    EXPECT_DOUBLE_EQ(position_.tilt, -0.3);
}

// ---- Combined move tests ----

TEST_F(MotionTest, CombinedMoveSetsBothAxes) {
    motion_->combined_move(1, -1, 0.01);  // pan right + tilt down

    const auto& calls = mock_->get_calls();
    // Should have at least 4 calls: set pan, set tilt, stop pan, stop tilt
    ASSERT_GE(calls.size(), 4u);

    // First two: set both speeds
    EXPECT_EQ(calls[0].first, CTRL_PAN_SPEED);
    EXPECT_EQ(calls[0].second, 1);
    EXPECT_EQ(calls[1].first, CTRL_TILT_SPEED);
    EXPECT_EQ(calls[1].second, -1);

    // Last two: stop both
    EXPECT_EQ(calls[2].first, CTRL_PAN_SPEED);
    EXPECT_EQ(calls[2].second, 0);
    EXPECT_EQ(calls[3].first, CTRL_TILT_SPEED);
    EXPECT_EQ(calls[3].second, 0);
}

TEST_F(MotionTest, CombinedMoveUpdatesBothPositionAxes) {
    position_.pan = 0.0;
    position_.tilt = 0.0;
    motion_->combined_move(1, 1, 0.2);

    EXPECT_DOUBLE_EQ(position_.pan, 0.2);
    EXPECT_DOUBLE_EQ(position_.tilt, 0.2);
}

// ---- Zoom absolute tests ----

TEST_F(MotionTest, ZoomAbsoluteSetsValue) {
    motion_->zoom_absolute(300);

    const auto& calls = mock_->get_calls();
    ASSERT_EQ(calls.size(), 1u);
    EXPECT_EQ(calls[0].first, CTRL_ZOOM_ABSOLUTE);
    EXPECT_EQ(calls[0].second, 300);
}

TEST_F(MotionTest, ZoomAbsoluteClampsAboveMax) {
    motion_->zoom_absolute(9999);

    const auto& calls = mock_->get_calls();
    ASSERT_EQ(calls.size(), 1u);
    EXPECT_EQ(calls[0].second, ZOOM_MAX);
    EXPECT_EQ(position_.zoom, ZOOM_MAX);
}

TEST_F(MotionTest, ZoomAbsoluteClampsBelowMin) {
    motion_->zoom_absolute(-50);

    const auto& calls = mock_->get_calls();
    ASSERT_EQ(calls.size(), 1u);
    EXPECT_EQ(calls[0].second, ZOOM_MIN);
    EXPECT_EQ(position_.zoom, ZOOM_MIN);
}

TEST_F(MotionTest, ZoomAbsoluteUpdatesPosition) {
    motion_->zoom_absolute(350);
    EXPECT_EQ(position_.zoom, 350);
}

// ---- Zoom relative tests ----

TEST_F(MotionTest, ZoomRelativeAddsToCurrentZoom) {
    // Position starts at ZOOM_DEFAULT (100)
    motion_->zoom_relative(50);

    EXPECT_EQ(position_.zoom, ZOOM_DEFAULT + 50);
}

TEST_F(MotionTest, ZoomRelativeClampsAboveMax) {
    position_.zoom = 480;
    motion_->zoom_relative(100);  // would be 580, clamped to 500

    EXPECT_EQ(position_.zoom, ZOOM_MAX);
}

TEST_F(MotionTest, ZoomRelativeClampsBelowMin) {
    position_.zoom = 120;
    motion_->zoom_relative(-200);  // would be -80, clamped to 100

    EXPECT_EQ(position_.zoom, ZOOM_MIN);
}

// ---- Stop test ----

TEST_F(MotionTest, StopSetsBothSpeedsToZero) {
    motion_->stop();

    const auto& calls = mock_->get_calls();
    ASSERT_EQ(calls.size(), 2u);
    EXPECT_EQ(calls[0].first, CTRL_PAN_SPEED);
    EXPECT_EQ(calls[0].second, 0);
    EXPECT_EQ(calls[1].first, CTRL_TILT_SPEED);
    EXPECT_EQ(calls[1].second, 0);
}

} // anonymous namespace
} // namespace bcc950

File: src/python/tests/unit/test_motion.py
"""Unit tests for MotionController."""

import threading
import pytest
from unittest.mock import MagicMock, call, patch

from bcc950.motion import MotionController
from bcc950.position import PositionTracker
from bcc950.v4l2_backend import V4L2Backend
from bcc950.constants import (
    CTRL_PAN_SPEED,
    CTRL_TILT_SPEED,
    CTRL_ZOOM_ABSOLUTE,
    ZOOM_MAX,
    ZOOM_MIN,
)


DEVICE = "/dev/video99"


@pytest.fixture
def backend():
    return MagicMock(spec=V4L2Backend)


@pytest.fixture
def position():
    return PositionTracker()


@pytest.fixture
def motion(backend, position):
    return MotionController(backend, DEVICE, position)


# ---------------------------------------------------------------------------
# Pan
# ---------------------------------------------------------------------------

class TestPan:
    def test_pan_right_duration(self, motion, backend):
        """Pan right: set speed +1, sleep, stop, update position."""
        with patch("bcc950.motion.time.sleep") as mock_sleep:
            motion.pan(1, duration=0.5)

        backend.set_control.assert_any_call(DEVICE, CTRL_PAN_SPEED, 1)
        backend.set_control.assert_any_call(DEVICE, CTRL_PAN_SPEED, 0)
        mock_sleep.assert_called_once_with(0.5)
        assert motion.position.pan == pytest.approx(0.5)

    def test_pan_left_duration(self, motion, backend):
        with patch("bcc950.motion.time.sleep") as mock_sleep:
            motion.pan(-1, duration=0.3)

        backend.set_control.assert_any_call(DEVICE, CTRL_PAN_SPEED, -1)
        backend.set_control.assert_any_call(DEVICE, CTRL_PAN_SPEED, 0)
        mock_sleep.assert_called_once_with(0.3)
        assert motion.position.pan == pytest.approx(-0.3)

    def test_pan_call_order(self, motion, backend):
        """set_control(speed) must happen before set_control(0)."""
        with patch("bcc950.motion.time.sleep"):
            motion.pan(1, duration=0.1)

        calls = backend.set_control.call_args_list
        assert calls[0] == call(DEVICE, CTRL_PAN_SPEED, 1)
        assert calls[1] == call(DEVICE, CTRL_PAN_SPEED, 0)


# ---------------------------------------------------------------------------
# Tilt
# ---------------------------------------------------------------------------

class TestTilt:
    def test_tilt_up_duration(self, motion, backend):
        with patch("bcc950.motion.time.sleep") as mock_sleep:
            motion.tilt(1, duration=0.4)

        backend.set_control.assert_any_call(DEVICE, CTRL_TILT_SPEED, 1)
        backend.set_control.assert_any_call(DEVICE, CTRL_TILT_SPEED, 0)
        mock_sleep.assert_called_once_with(0.4)
        assert motion.position.tilt == pytest.approx(0.4)

    def test_tilt_down_duration(self, motion, backend):
        with patch("bcc950.motion.time.sleep") as mock_sleep:
            motion.tilt(-1, duration=0.2)

        backend.set_control.assert_any_call(DEVICE, CTRL_TILT_SPEED, -1)
        backend.set_control.assert_any_call(DEVICE, CTRL_TILT_SPEED, 0)
        mock_sleep.assert_called_once_with(0.2)
        assert motion.position.tilt == pytest.approx(-0.2)


# ---------------------------------------------------------------------------
# Zoom
# ---------------------------------------------------------------------------

class TestZoom:
    def test_zoom_absolute_normal(self, motion, backend):
        motion.zoom_absolute(250)
        backend.set_control.assert_called_once_with(DEVICE, CTRL_ZOOM_ABSOLUTE, 250)
        assert motion.position.zoom == 250

    def test_zoom_absolute_clamps_low(self, motion, backend):
        motion.zoom_absolute(0)
        backend.set_control.assert_called_once_with(DEVICE, CTRL_ZOOM_ABSOLUTE, ZOOM_MIN)
        assert motion.position.zoom == ZOOM_MIN

    def test_zoom_absolute_clamps_high(self, motion, backend):
        motion.zoom_absolute(9999)
        backend.set_control.assert_called_once_with(DEVICE, CTRL_ZOOM_ABSOLUTE, ZOOM_MAX)
        assert motion.position.zoom == ZOOM_MAX

    def test_zoom_relative_positive(self, motion, backend):
        motion.zoom_relative(50)
        backend.set_control.assert_called_once_with(DEVICE, CTRL_ZOOM_ABSOLUTE, ZOOM_MIN + 50)
        assert motion.position.zoom == ZOOM_MIN + 50

    def test_zoom_relative_clamps_at_min(self, motion, backend):
        """Negative delta from minimum should stay at ZOOM_MIN."""
        motion.zoom_relative(-50)
        backend.set_control.assert_called_once_with(DEVICE, CTRL_ZOOM_ABSOLUTE, ZOOM_MIN)
        assert motion.position.zoom == ZOOM_MIN

    def test_zoom_relative_clamps_at_max(self, motion, backend):
        motion.position.zoom = ZOOM_MAX
        motion.zoom_relative(100)
        backend.set_control.assert_called_once_with(DEVICE, CTRL_ZOOM_ABSOLUTE, ZOOM_MAX)
        assert motion.position.zoom == ZOOM_MAX


# ---------------------------------------------------------------------------
# Combined moves
# ---------------------------------------------------------------------------

class TestCombinedMove:
    def test_combined_move_sets_and_stops_both(self, motion, backend):
        with patch("bcc950.motion.time.sleep"):
            motion.combined_move(1, -1, duration=0.2)

        calls = backend.set_control.call_args_list
        # First two calls: set speeds
        assert calls[0] == call(DEVICE, CTRL_PAN_SPEED, 1)
        assert calls[1] == call(DEVICE, CTRL_TILT_SPEED, -1)
        # After sleep: stop both
        assert calls[2] == call(DEVICE, CTRL_PAN_SPEED, 0)
        assert calls[3] == call(DEVICE, CTRL_TILT_SPEED, 0)

    def test_combined_move_updates_position(self, motion, backend):
        with patch("bcc950.motion.time.sleep"):
            motion.combined_move(1, -1, duration=0.5)

        assert motion.position.pan == pytest.approx(0.5)
        assert motion.position.tilt == pytest.approx(-0.5)

    def test_combined_move_with_zoom(self, motion, backend):
        with patch("bcc950.motion.time.sleep"):
            motion.combined_move_with_zoom(1, 1, 300, duration=0.2)

        calls = backend.set_control.call_args_list
        assert call(DEVICE, CTRL_PAN_SPEED, 1) in calls
        assert call(DEVICE, CTRL_TILT_SPEED, 1) in calls
        assert call(DEVICE, CTRL_ZOOM_ABSOLUTE, 300) in calls
        # Stops
        assert call(DEVICE, CTRL_PAN_SPEED, 0) in calls
        assert call(DEVICE, CTRL_TILT_SPEED, 0) in calls
        # Position updated
        assert motion.position.pan == pytest.approx(0.2)
        assert motion.position.tilt == pytest.approx(0.2)
        assert motion.position.zoom == 300

    def test_combined_move_with_zoom_clamps(self, motion, backend):
        """Zoom target above ZOOM_MAX should be clamped."""
        with patch("bcc950.motion.time.sleep"):
            motion.combined_move_with_zoom(0, 0, 9999, duration=0.1)

        backend.set_control.assert_any_call(DEVICE, CTRL_ZOOM_ABSOLUTE, ZOOM_MAX)
        assert motion.position.zoom == ZOOM_MAX


# ---------------------------------------------------------------------------
# Thread safety
# ---------------------------------------------------------------------------

class TestThreadSafety:
    def test_concurrent_moves_do_not_interleave(self, motion, backend):
        """Two concurrent pan calls should each complete their start-stop
        sequence without interleaving (ensured by the internal lock).

        We verify that set_control calls always appear in paired
        start/stop sequences -- never two starts in a row without a stop.
        """
        results = []
        barrier = threading.Barrier(2)

        def _pan_right():
            barrier.wait()
            with patch("bcc950.motion.time.sleep"):
                motion.pan(1, duration=0.05)

        def _pan_left():
            barrier.wait()
            with patch("bcc950.motion.time.sleep"):
                motion.pan(-1, duration=0.05)

        t1 = threading.Thread(target=_pan_right)
        t2 = threading.Thread(target=_pan_left)
        t1.start()
        t2.start()
        t1.join(timeout=5)
        t2.join(timeout=5)

        calls = backend.set_control.call_args_list
        pan_calls = [c for c in calls if c[0][1] == CTRL_PAN_SPEED]

        # Should have exactly 4 pan calls: start-stop, start-stop
        assert len(pan_calls) == 4

        # Verify pairing: calls at index 0,1 form a pair, 2,3 form a pair
        # First of each pair is non-zero, second is zero
        assert pan_calls[0][0][2] != 0  # first start
        assert pan_calls[1][0][2] == 0  # first stop
        assert pan_calls[2][0][2] != 0  # second start
        assert pan_calls[3][0][2] == 0  # second stop


# ---------------------------------------------------------------------------
# Stop
# ---------------------------------------------------------------------------

class TestStop:
    def test_stop_zeros_both_axes(self, motion, backend):
        motion.stop()
        calls = backend.set_control.call_args_list
        assert call(DEVICE, CTRL_PAN_SPEED, 0) in calls
        assert call(DEVICE, CTRL_TILT_SPEED, 0) in calls
